File: src/main/java/travelator/Tracking.java
package travelator;

import java.util.Optional;

import static java.util.stream.Collectors.toList;
import static travelator.Trip.BookingStatus.BOOKED;

class Tracking implements ITrackTrips {

    private final Trips trips;

    public Tracking(Trips trips) {
        this.trips = trips;
    }

    @Override
    public Optional<Trip> currentTripFor(String customerId) {
        var candidates = trips.currentTripsFor(customerId).stream()
            .filter((trip) -> trip.getBookingStatus() == BOOKED)
            .collect(toList());
        if (candidates.size() == 1)
            return Optional.of(candidates.get(0));
        else if (candidates.size() == 0)
            return Optional.empty();
        else
            throw new IllegalStateException(
                "Unexpectedly more than one current trip for " + customerId
            );
    }
}

File: src/main/java/travelator/Tracking.kt
package travelator

import travelator.Trip.BookingStatus
import java.time.Instant

class Tracking(
    private val trips: Trips
) : ITrackTrips {

    override fun currentTripFor(customerId: String, at: Instant): Trip? {
        val candidates = trips.currentTripsFor(customerId, at)
            .filter { it.bookingStatus == BookingStatus.BOOKED }
        return when (candidates.size) {
            1 -> candidates[0]
            0 -> null
            else -> throw IllegalStateException(
                "Unexpectedly more than one current trip for $customerId"
            )
        }
    }
}